### ISSUE TITLE
RFC - Use Association finder internally

### ISSUE THREAD

**Issue by markstory** [cakephp/cakephp#12881]
This is a (multiple allowed):

* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 4.0
* Platform and Target: N/A

Presently Associations have both `conditions` and `finder` options that enable custom logic to be applied when associations retrieve data. When doing update or delete operations only `conditions` is applied. This makes using `finder` impractical as you need to duplicate all association conditions between the finder and the association `conditions` option.

To reduce duplication and make the ORM simpler to use we should support the `finder` option for all Association proxy methods. The following methods would need to change:

* Association::exists()
* Association::updateAll()
* Association::deleteAll()

All of these operations should be updated so that they all find() first, then apply any association conditions. We would also be able to deprecate association `conditions` option without a warning and in a future 4.x release a warning could be added. This would make `finder` the canoncial way to create scoped associations.

I'm proposing this change for 4.0 as it requires changing a number of ORM internals which could be risky.

### Related issues

* #12859

**Comment by lorenzo** [comment]
I agree with this change.

**Comment by saeideng** [comment]
>We would also be able to deprecate association conditions option

:+1:  if we able to create a query/finder inline 
```php
$this->belongsTo('Authors', [
            //
            ])->setXYZ('...')
              ->setFinder(function(Query $query, array $options){
                     $query->where(['published' => true]);
             })
```

**Comment by markstory** [comment]
I implemented this in #12929

**Comment by ADmad** [comment]
Closing as related PR has been merged.
